### PR TITLE
FE-1396 - Ejected create-react-app tools-ui upgrade

### PR DESCRIPTION
**Pull request by nstanard**
Local: http://localhost:3000/dkim
Stage: https://tools-staging.sparkpost.com/dkim
Prod: https://tools.sparkpost.com/dkim

### What Changed
 - Re-initialized the project with `create-react-app`, then ejected the starter. Then added the src and config files back.
 - Cypress was installed and a few basic tests were added for each of the three pages.
 - The local application no longer redirects to production. [Instead, it redirects to localhost](https://github.com/SparkPost/tools-ui/pull/76/files#diff-481bff25593155cc0695aeb0f112c109e3075ef703d254734ec2a3c5266ee0beR14).  (`npm run build` will still redirect to prod while `npm run start` redirects to localhost)
 - `states/state-max-refresh-retries.json` was deleted.
 - `config/polyfills.js` was deleted.
 - React was upgraded (via create-react-app)
 - WebPack was upgraded (via create-react-app)
 - redux-form was upgraded.

### How to test
 - Run this project locally while having the stage or production app open too.
 - There are three main tools/routes for this app. The DKIM Validator, SPF Inspector, and SPF Builder.
 - We need to make sure that every piece of functionality on all three of those pages/states works exactly as expected and like production. We should be on the lookout for javascript (mostly type) errors especially. 
 - 
 

#### Navigation
 - Tools icon/button/link goes to the marketing landing page.
 - Login button -> redirects to 2web2ui login (works with local 2web2ui with the right configs)
 - Signup button -> redirects to 2web2ui signup (works with local 2web2ui with the right configs)

#### DKIM Validator Feature List
 - Generate email address
 - View Results takes user to the result view/state (Requires sending an email to the generated address) 
 - **What do we need to send to that email to accurately test the DKIM View Results page?**

#### SPF Inspector Feature List
 - Inspect a domain 
 - View domain inspection results
 - Inspect results: Save Results
 - Inspect results: Share Results
 - Inspect results: Refresh
 - Inspect results: How do I fix errors?
 - Inspect results: Expand / Collapse
 - Inspect results: Shows SPF Record and IP list

#### SPF Builder 
 - Enter a domain
 - MX Records
 - A Records
 - IP Network Ranges
 - Includes
 - How strict should this SPF Record be?
 - Fixed view "Add this TXT record to your DNS" message
 - "check out the inspector" link inside the fixed view message.
 - Copy the generated string/record from adding the different configurations.

### CODE REVIEW

File: src/config/default.js [hunk 1]
```diff
@@ -11,7 +12,7 @@ export default {
       path: '/'
     },
     requestHeaders: {
-      authHeader: 'Basic bXN5c1VJTGltaXRlZDphZjE0OTdkYS02NjI5LTQ3NTEtODljZS01ZDBmODE4N2MyMDQ=',
+      Authorization: 'Basic bXN5c1VJTGltaXRlZDphZjE0OTdkYS02NjI5LTQ3NTEtODljZS01ZDBmODE4N2MyMDQ=',
```
Comment (nstanard) [inline src/config/default.js:15]: Lets talk about adding this as a travis ENV variable - and have developers set this locally as an ENV variable too. (down the road, not on this PR)

Webpack variable replace (DefinePlugin? the env variable that gets set)

**Comment by nstanard** [comment]
Current Login button click on master: 
image
`http://app.sparkpost.test/auth?return=http%3A%2F%2Flocalhost%3A3000%2Fspf%2Fbuilder&src=SP-Tools&sfdcid=701600000019BWh`


This branch Login button click: 
image
`http://app.sparkpost.test/auth?return=http%3A%2F%2Flocalhost%3A3000%2Fdkim&src=SP-Tools&sfdcid=701600000019BWh`

File: src/routes.js [hunk 2]
```diff
@@ -11,7 +11,11 @@ import SPFBuilder from 'pages/builder/Builder';
 
 const redirectRoot = () => {
   const { search } = browserHistory.getCurrentLocation();
-  window.location = `https://www.sparkpost.com/email-tools${search}`;
+  if (process.env.NODE_ENV === 'development') {
```
Comment (nstanard) [inline src/routes.js:14]: This was a choice I made, wasn't 100% needed but I feel it's right.

File: config/env.js [hunk 3]
```diff
@@ -1,32 +1,110 @@
...
+  .split(path.delimiter)
+  .filter(folder => folder && !path.isAbsolute(folder))
+  .map(folder => path.resolve(appDirectory, folder))
+  .join(path.delimiter);
+
+// Grab NODE_ENV and REACT_APP_* environment variables and prepare them to be
+// injected into the application via DefinePlugin in webpack configuration.
+const REACT_APP = /^REACT_APP_/i;
 
 // Default REACT_APP_ENV to 'development' if not already set
```
Comment (nstanard) [inline config/env.js:63]: This is an example of something the original team customized from the build. I had to add this back. I'm on the lookout for more things like this.

File: src/pages/builder/containers/FormContainer.js [hunk 4]
```diff
@@ -10,6 +10,9 @@ import IpRangesContainer from '../containers/IpRangesContainer';
 class FormContainer extends Component {
   render() {
     const { form } = this.props;
+    if (!form || !form.values) {
```
Comment (nstanard) [inline src/pages/builder/containers/FormContainer.js:13]: Need feedback on this decision, the form is sometimes coming through undefined.

Comment (nstanard) [reply src/pages/builder/containers/FormContainer.js:13]: For testing and verifying - we need to make sure we are using redux dev tools and looking at the redux-form events/cycles.
image

File: README.md [hunk 5]
```diff
@@ -45,3 +41,15 @@ For releasing to production, other than merging to `master`, you need to tag it.
 
 - `npm version <major|minor|patch>`. Use appropriate option among `major|minor|patch`. A new tag will be created.
 - `git push && git push --tags`
+
+Note: If you add the following to your `.gitconfig` file, you can just push and tags will always be pushed too.
```
Comment (nstanard) [inline README.md:45]: Incorporated some feedback/info from: https://github.com/SparkPost/tools-ui/pull/44

File: src/pages/spf/QueryPage.js [hunk 6]
```diff
@@ -52,7 +52,8 @@ export class QueryPage extends Component {
 
           <div className='col-xs-12 col-md-8'>
             <div className={classNames('input__group', {'has-error': this.state.domainError })}>
-              <input className='input__text input--full' type='text' onChange={(e) => this.setState({domain: e.target.value}) } placeholder='mydomain.com' />
+              <label className="inspect-domain-label" for="inspect-domain">Inspect Domain</label>
```
Comment (nstanard) [inline src/pages/spf/QueryPage.js:55]: image
currently not looking good with this label - fixing now

File: config/polyfills.js [hunk 7]
```diff
@@ -1,14 +0,0 @@
-if (typeof Promise === 'undefined') {
```
Comment (nstanard) [inline config/polyfills.js:1]: I wonder if we need these anymore - I didn't add them back after ejecting create-react-app.

File: src/config/default.js [hunk 8]
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line import/no-anonymous-default-export
```
Comment (nstanard) [inline src/config/default.js:1]: A lot of eslint warnings/errors showed up that I disabled (for now) - if we ever revisit this project we can fix those one at a time.

File: src/pages/spf/QueryPage.js (hunk 6 above)
Comment (nstanard) [reply src/pages/spf/QueryPage.js:55]: fixed
image

File: src/helpers/http.js [hunk 9]
```diff
@@ -13,10 +13,7 @@ function useRefreshToken(refresh_token) {
     method: 'POST',
     url: '/authenticate',
     data: `grant_type=refresh_token&refresh_token=${refresh_token}`,
-    headers: {
```
Comment (nstanard) [inline src/helpers/http.js:16]: 🤔  I didn't change this - researching now.

File: cypress/integration/DKIM/dkim.spec.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  it('has the correct page title.', () => {
+    cy.title().should('include', 'DKIM Validator');
+  });
+
+  it('has the correct page header.', () => {
+    cy.findByRole('heading', { name: 'DKIM Validator' }).should('be.visible');
+  });
+
+  it('generates an email address.', () => {
+    // cy.intercept({
```
Comment (nstanard) [inline cypress/integration/DKIM/dkim.spec.js:19]: cy. is currently being flagged as undefined but I added the cypress plugin to the config so I need to double check this at some point.

image

File: src/routes.js (hunk 2 above)
Comment (PatrickSison) [reply src/routes.js:14]: I think I'm still getting redirected locally. I think we need to modify package.json for this

File: src/pages/spf/QueryPage.js (hunk 6 above)
Comment (PatrickSison) [inline src/pages/spf/QueryPage.js:55]: 👍

File: package.json [hunk 11]
```diff
@@ -1,137 +1,184 @@
...
+    "cypress": "^6.8.0",
+    "eslint-plugin-cypress": "^2.11.2",
+    "eslint-plugin-prettier": "^3.3.1",
+    "react-test-renderer": "^17.0.1",
+    "serve": "^11.3.2"
   },
   "scripts": {
-    "start": "npm run fix && node scripts/start.js",
+    "serve": "serve -s build",
+    "start": "node scripts/start.js",
```
Comment (PatrickSison) [inline package.json:99]: Should this set the environment variable for development?

File: package.json [hunk 12]
```diff
@@ -1,137 +1,184 @@
...
     "build": "node scripts/build.js",
-    "test": "npm run fix && node scripts/test.js --env=jsdom",
-    "lint": "eslint \"src/**/*.js\"",
-    "fix": "npm run lint -- --fix"
+    "test": "node scripts/test.js",
+    "lint": "eslint 'src/**/*.js'",
+    "fix": "npm run lint -- --fix",
+    "cy": "cross-env NODE_ENV=test cypress open --config-file cypress.dev.json"
```
Comment (PatrickSison) [inline package.json:104]: Should we include a separate environment variable to run this against a locally running version of this app?

**Review by PatrickSison — COMMENTED** [review]
A couple things so far. I've been trying to troubleshoot cypress running as well on my local environment. The changes seem good for the most part right now

File: package.json (hunk 11 above)
Comment (PatrickSison) [reply package.json:99]: Nevermind, see it in the build script right now. I'm still getting redirected to prod

File: package.json (hunk 12 above)
Comment (nstanard) [reply package.json:104]: That's a good question, I might try to take off the NODE_ENV from this line to see if it matters. I wasn't sure tbh. But for now I was only planning on running these tests locally and not in stage/prod/travis.

**Comment by nstanard** [comment]
One callout I'm noticing while testing. The production app right now doesn't have the "Delete my email cookie and start over" button and I"m not sure why. 

Production:
image

This branch/local:
image

File: config/polyfills.js (hunk 7 above)
Comment (nstanard) [reply config/polyfills.js:1]: This file was originally from the ejection of create-react-app and so my thought is it's no longer needed:
image

(the new create-react-app ejection did not output this file)

File: package.json [hunk 13]
```diff
@@ -1,137 +1,184 @@
...
+    "file-loader": "6.1.1",
+    "fs-extra": "^9.0.1",
+    "html-webpack-plugin": "4.5.0",
+    "identity-obj-proxy": "3.0.0",
+    "jest": "26.6.0",
+    "jest-circus": "26.6.0",
+    "jest-resolve": "26.6.0",
+    "jest-watch-typeahead": "0.6.1",
     "js-cookie": "^2.1.3",
-    "lodash": "^4.17.4",
```
Comment (nstanard) [inline package.json:66]: need to double check why lodash isn't being installed now - everything seems to be working though... so need to double check this

File: src/index.js [hunk 14]
```diff
@@ -22,3 +26,8 @@ ReactDOM.render((
     <Router history={browserHistory} routes={routes} render={applyRouterMiddleware(useScroll())} />
   </Provider>
 ), document.getElementById('root'));
+
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals();
```
Comment (nstanard) [inline src/index.js:33]: we should possibly remove the reportWebVitals() - it came from create-react-app

File: .node-version [hunk 15]
```diff
@@ -1 +1 @@
-10.23.1
```
Comment (nstanard) [inline .node-version:1]: Set this as the same version travis uses

image

File: package.json (hunk 13 above)
Comment (nstanard) [reply package.json:66]: image
Added this back just to be safe - it's used in a number of spots

**Comment by nstanard** [comment]
Closing in favor of the un-ejected tools-ui

**Pull request closed without merging**